docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,342 @@
+
+* The end of the message.
+* The end of the received buffer if no message was found.
+
+The single message case has the most potential for errors. Passing the wrong values to *examined* can result in an out of memory exception or an infinite loop. For more information, see the [PipeReader common problems](#gotchas) section in this article.
+
+### Reading multiple messages
+
+The following code reads all messages from a `PipeReader` and calls `ProcessMessageAsync` on each.
+
+[!code-csharp[MyConnection1](~/samples/snippets/csharp/pipelines/MyConnection1.cs?name=snippet)]
+
+### Cancellation
+
+`PipeReader.ReadAsync`:
+
+* Supports passing a <xref:System.Threading.CancellationToken>.
+* Throws in an <xref:System.OperationCanceledException> if the `CancellationToken` is canceled while there's a read pending.
+* Supports a way to cancel the current read operation via <xref:System.IO.Pipelines.PipeReader.CancelPendingRead%2A?displayProperty=nameWithType>, which avoids raising an exception. Calling `PipeReader.CancelPendingRead` returns a <xref:System.IO.Pipelines.ReadResult> with `IsCanceled` set to `true`. This can be useful for halting the existing read loop in a non-destructive and non-exceptional way.
+
+[!code-csharp[MyConnection](~/samples/snippets/csharp/pipelines/MyConnection.cs?name=snippet)]
+
+<a name="gotchas"></a>
+
+### PipeReader common problems
+
+* Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed.
+* Passing `buffer.End` as examined may result in:
+
+  * Stalled data
+  * Possibly an eventual Out of Memory (OOM) exception if data isn't consumed. For example, `PipeReader.AdvanceTo(position, buffer.End)` when processing a single message at a time from the buffer.
+
+* Passing the wrong values to `consumed` or `examined` may result in an infinite loop. For example, `PipeReader.AdvanceTo(buffer.Start)` if `buffer.Start` hasn't changed will cause the next call to `PipeReader.ReadAsync` to return immediately before new data arrives.
+* Passing the wrong values to `consumed` or `examined` may result in infinite buffering (eventual OOM). For example, `PipeReader.AdvanceTo(buffer.Start, buffer.End)` unconditionally when processing a single message at a time from the buffer.
+* Using the `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` may result in memory corruption (use after free).
+* Failing to call `PipeReader.Complete/CompleteAsync` may result in a memory leak.
+* Checking <xref:System.IO.Pipelines.ReadResult.IsCompleted?displayProperty=nameWithType> and exiting the reading logic before processing the buffer results in data loss. The loop exit condition should be based on `ReadResult.Buffer.IsEmpty` and `ReadResult.IsCompleted`. Doing this in the wrong order could result in an infinite loop.
```

Comment:
Data loss I get. But how does checking IsEmpty and IsCompleted in the wrong order lead to an infinite loop?

Reply:
I see after looking at the "Infinite loop" problematic code. It's not checking them in the wrong order that's problematic, it's expecting *both* to be true before you exit that can lead to an infinite loop given a stream ends with a partial message.

Reply:
```suggestion
* Checking <xref:System.IO.Pipelines.ReadResult.IsCompleted?displayProperty=nameWithType> and exiting the reading logic before processing the buffer results in data loss. The loop exit condition should be based on `ReadResult.Buffer.IsEmpty` and `ReadResult.IsCompleted`. Doing this incorrectly could result in an infinite loop.
```

Let's just go with that.

Reply:
> Which sentence is that?

That quote was from [another comment](https://github.com/dotnet/docs/pull/14414#discussion_r333307312), but
I was referring to "For example, passing a `SequencePosition` that was already processed."

You applied the edit I suggested which improved it slightly, but I think that even the new expanded version is unclear and unnecessary. I'll suggest another edit to just remove it.